tools/lib/bpf/libbpf.c

```diff
@@ -6144,8 +6130,50 @@ bpf_object__relocate(struct bpf_object *obj, const char *targ_btf_path)
 			return err;
 		}
 	}
-	if (!obj->gen_loader)
-		bpf_object__free_relocs(obj);
+
+	return 0;
+}
+
+/* relocate instructions that refer to map fds */
+static int
+bpf_object__finish_relocate(struct bpf_object *obj)
+{
+	int i, j;
+
+	if (obj->gen_loader)
+		return 0;
+
+	for (i = 0; i < obj->nr_programs; i++) {
+		struct bpf_program *prog = &obj->programs[i];
+
+		if (prog_is_subprog(obj, prog))
+			continue;
+		for (j = 0; j < prog->nr_reloc; j++) {
+			struct reloc_desc *relo = &prog->reloc_desc[j];
+			struct bpf_insn *insn = &prog->insns[relo->insn_idx];
+			struct extern_desc *ext;
+
+			switch (relo->type) {
+			case RELO_LD64:
+				insn[0].src_reg = BPF_PSEUDO_MAP_FD;
+				insn[0].imm = obj->maps[relo->map_idx].fd;
+				break;
+			case RELO_DATA:
+				insn[0].src_reg = BPF_PSEUDO_MAP_VALUE;
+				insn[0].imm = obj->maps[relo->map_idx].fd;
+				break;
+			case RELO_EXTERN_VAR:
+				ext = &obj->externs[relo->sym_off];
+				if (ext->type == EXT_KCFG) {
+					insn[0].src_reg = BPF_PSEUDO_MAP_VALUE;
+					insn[0].imm = obj->maps[obj->kconfig_map_idx].fd;
```

Comment:
my linter complains on the lack of break; here (I know it would fall to "default" break). Should we add break here to satisfy linters ?

Reply:
Good catch. Added it.